File: sdk/typescript/src/api/registry.ts
import type { SigningKey } from "../auth.js";
import { signFreshCanonicalPayload } from "../auth.js";
import { canonicalPayload } from "../crypto.js";
import {
  TinyVerseError,
  type HttpClient,
  type PaymentChallenge,
} from "../http.js";
import {
  buildX402PaymentMap,
  type X402PaymentMap,
  type X402PaymentMapOptions,
} from "../x402.js";
import {
  executeSolanaX402Payment,
  SOLANA_MAINNET_NETWORK,
  SOLANA_USDC_MINT,
  type SolanaX402PaymentExecution,
  type SolanaX402PaymentExecutionOptions,
} from "../solana.js";
import type {
  ActorType,
  AvailabilityResponse,
  Identity,
  IdentityClaimRequest,
  IdentityExport,
  IdentityTransferRequest,
  PaymentMethod,
  ProfileVisibility,
  ProfileVisibilityUpdate,
  RenewalRequest,
  Subname,
  SubnameCreateRequest,
} from "../types/index.js";

export interface RegisterRequest {
  username: string;
  cryptoId: string;
  publicKey: string;
  paymentMethods?: Array<PaymentMethod>;
  /**
   * The wallet's self-declared, trust-based actor type ("human"/"agent"),
   * recorded on the wallet's User profile when it is first provisioned. Not part
   * of the signed payload — the backend trusts the claim. Defaults to "agent".
   */
  actorType?: ActorType;
  /**
   * Request that this name be assigned as the wallet's primary handle. When
   * omitted, the backend still auto-assigns it as primary if the wallet has no
   * primary yet (a wallet's first name is always its primary).
   */
  primary?: boolean;
  payment?: Record<string, string>;
  signature?: string;
}

export interface SolanaRegistrationPaymentOptions
  extends Omit<SolanaX402PaymentExecutionOptions, "payment" | "signer"> {
  amount?: string;
  to?: string;
  asset?: string;
  network?: string;
  nonce?: string;
  expiresAt?: string;
  expiresInMs?: number;
  metadata?: Record<string, string>;
  registrationAttempts?: number;
  registrationIntervalMs?: number;
  registrationRetryErrors?: Array<string>;
}

export interface SolanaRegistrationResult {
  identity: Identity;
  payment: SolanaX402PaymentExecution;
}

export interface SolanaRegistrationFailure extends Error {
  registrationPayment?: SolanaX402PaymentExecution | X402PaymentMap;
  onChainTx?: string;
}

export interface SolanaRegistrationProofOptions
  extends Partial<Pick<
    X402PaymentMapOptions,
    "amount" | "asset" | "network" | "nonce" | "expiresAt" | "expiresInMs" | "metadata"
  >> {
  onChainTx: string;
  to?: string;
  registrationAttempts?: number;
  registrationIntervalMs?: number;
  registrationRetryErrors?: Array<string>;
}

export interface SolanaRegistrationProofResult {
  identity: Identity;
  onChainTx: string;
  payment: X402PaymentMap;
}

const DEFAULT_REGISTRATION_ATTEMPTS = 30;
const DEFAULT_REGISTRATION_INTERVAL_MS = 3000;
const DEFAULT_REGISTRATION_RETRY_ERRORS = [
  "transaction not found",
  "insufficient confirmations",
];

export class RegistryApi {
  constructor(
    private readonly http: HttpClient,
    private readonly signingKey?: SigningKey,
  ) {}

  async register(request: RegisterRequest): Promise<Identity> {
    request = {
      ...request,
      username: normalizeHandle(request.username),
    };

    let signature: string | undefined;
    if (this.signingKey && !request.signature) {
      const payload = registrationSignaturePayload(request);
      signature = await signFreshCanonicalPayload(this.signingKey, payload);
    }

    return this.http.postPublic<Identity>("/registry/names", {
      ...request,
      ...(signature ? { signature } : {}),
    });
  }

  async registerWithSolanaPayment(
    request: Omit<RegisterRequest, "payment"> & { payment?: never },
    options: SolanaRegistrationPaymentOptions,
  ): Promise<SolanaRegistrationResult> {
    if (!this.signingKey) {
      throw new Error("registerWithSolanaPayment requires a signing key");
    }

    const normalizedRequest: RegisterRequest = {
      ...request,
      username: normalizeHandle(request.username),
    };
    const challenge =
      options.amount && options.to
        ? undefined
        : await this.registrationPaymentChallenge(normalizedRequest);
    const amount = options.amount ?? challenge?.amount;
    const to = options.to ?? challenge?.to;
    if (!amount || !to) {
      throw new Error("registration payment requires amount and recipient");
    }

    const payment = await executeSolanaX402Payment({
      ...options,
      mint: options.mint ?? SOLANA_USDC_MINT,
      signer: this.signingKey,
      payment: {
        scheme: "exact",
        network: options.network ?? challenge?.network ?? SOLANA_MAINNET_NETWORK,
        asset: options.asset ?? challenge?.asset ?? "USDC",
        amount,
        from: normalizedRequest.cryptoId,
        to,
        nonce:
          options.nonce ??
          challenge?.nonce ??
          generateRegistrationNonce(normalizedRequest.username),
        expiresAt: options.expiresAt ?? challenge?.expiresAt,
        expiresInMs: options.expiresInMs,
        metadata: {
          ...challenge?.metadata,
          identity: normalizedRequest.username,
          purpose: "registration",
          ...options.metadata,
        },
        publicKeyBase64: normalizedRequest.publicKey,
      },
    });
    let identity: Identity;
    try {
      identity = await this.registerWithPaymentMap(
        normalizedRequest,
        payment.payment,
        options,
      );
    } catch (error) {
      throw attachRegistrationPayment(error, payment);
    }

    return { identity, payment };
  }

  async registerWithExistingSolanaPayment(
    request: Omit<RegisterRequest, "payment"> & { payment?: never },
    options: SolanaRegistrationProofOptions,
  ): Promise<SolanaRegistrationProofResult> {
    if (!this.signingKey) {
      throw new Error("registerWithExistingSolanaPayment requires a signing key");
    }

    const normalizedRequest: RegisterRequest = {
      ...request,
      username: normalizeHandle(request.username),
    };
    const challenge =
      options.amount && options.to
        ? undefined
        : await this.registrationPaymentChallenge(normalizedRequest);
    const amount = options.amount ?? challenge?.amount;
    const to = options.to ?? challenge?.to;
    if (!amount || !to) {
      throw new Error("registration payment requires amount and recipient");
    }

    const payment = await buildX402PaymentMap(this.signingKey, {
      scheme: "exact",
      network: options.network ?? challenge?.network ?? SOLANA_MAINNET_NETWORK,
      asset: options.asset ?? challenge?.asset ?? "USDC",
      amount,
      from: normalizedRequest.cryptoId,
      to,
      nonce:
        options.nonce ??
        challenge?.nonce ??
        generateRegistrationNonce(normalizedRequest.username),
      expiresAt: options.expiresAt ?? challenge?.expiresAt,
      expiresInMs: options.expiresInMs,
      onChainTx: options.onChainTx,
      tx: options.onChainTx,
      transaction: options.onChainTx,
      metadata: {
        ...challenge?.metadata,
        identity: normalizedRequest.username,
        purpose: "registration",
        ...options.metadata,
      },
      publicKeyBase64: normalizedRequest.publicKey,
    });
    let identity: Identity;
    try {
      identity = await this.registerWithPaymentMap(
        normalizedRequest,
        payment,
        options,
      );
    } catch (error) {
      throw attachRegistrationPayment(error, payment);
    }

    return { identity, onChainTx: options.onChainTx, payment };
  }

  private async registrationPaymentChallenge(
    request: RegisterRequest,
  ): Promise<PaymentChallenge | undefined> {
    try {
      await this.register(request);
    } catch (error) {
      if (error instanceof TinyVerseError && error.status === 402) {
        return error.paymentRequired?.payment;
      }
      throw error;
    }
    throw new Error("registration did not return a payment challenge");
  }

  private async registerWithPaymentMap(
    request: RegisterRequest,
    payment: X402PaymentMap,
    options: {
      registrationAttempts?: number;
      registrationIntervalMs?: number;
      registrationRetryErrors?: Array<string>;
    },
  ): Promise<Identity> {
    try {
      return await this.registerRetryingPayment({
        ...request,
        payment,
      }, {
        attempts: options.registrationAttempts,
        intervalMs: options.registrationIntervalMs,
        retryErrors: options.registrationRetryErrors,
      });
    } catch (error) {
      const recovered = await this.createdIdentityAfterRegistrationError(
        request.username,
        error,
      );
      if (!recovered) {
        throw error;
      }
      return recovered;
    }
  }

  private async createdIdentityAfterRegistrationError(
    username: string,
    error: unknown,
  ): Promise<Identity | undefined> {
    if (!(error instanceof TinyVerseError) || error.status < 500) {
      return undefined;
    }
    try {
      const result = await this.get(username);
      return result.available ? undefined : result.identity;
    } catch {
      return undefined;
    }
  }

  get(name: string): Promise<AvailabilityResponse> {
    return this.http.get<AvailabilityResponse>(
      `/registry/names/${encodeURIComponent(name)}`,
    );
  }

  export(name: string): Promise<IdentityExport> {
    return this.http.get<IdentityExport>(
      `/registry/names/${encodeURIComponent(name)}/export`,
    );
  }

  async updateProfileVisibility(
    name: string,
    update: ProfileVisibilityUpdate,
  ): Promise<ProfileVisibility> {
    if (this.signingKey && !update.signature) {
      const payload = canonicalPayload("identity.profile.visibility", {
        activity: update.activity ?? null,
        agentCard: update.agentCard ?? null,
        attestations: update.attestations ?? null,
        broadcasts: update.broadcasts ?? null,
        groups: update.groups ?? null,
        searchEngineIndexing: update.searchEngineIndexing ?? null,
        username: name,
      });
      update = {
        ...update,
        signature: await signFreshCanonicalPayload(this.signingKey, payload),
      };
    }
    return this.http.putDirectoryAuth<ProfileVisibility>(
      `/registry/names/${encodeURIComponent(name)}/profile-visibility`,
      update,
    );
  }

  async renew(name: string, request: RenewalRequest): Promise<Identity> {
    if (this.signingKey && !request.signature) {
      const payload = canonicalPayload("identity.renew", { username: name });
      request = {
        ...request,
        signature: await signFreshCanonicalPayload(this.signingKey, payload),
      };
    }
    return this.http.postDirectoryAuth<Identity>(
      `/registry/names/${encodeURIComponent(name)}/renew`,
      request,
    );
  }

  /**
   * Directly transfer this name to another wallet with no payment (a gift or
   * account move), distinct from the paid marketplace flow. The CURRENT owner
   * (this client's signing key, or an approved delegate) authorizes the move;
   * `request.cryptoId`/`request.publicKey` identify the recipient. The name
   * keeps its registration period and the recipient receives it unassigned.
   */
  async transfer(
    name: string,
    request: IdentityTransferRequest,
  ): Promise<Identity> {
    if (this.signingKey && !request.signature) {
      const payload = canonicalPayload("identity.transfer", {
        cryptoId: request.cryptoId,
        publicKey: request.publicKey,
        username: name,
      });
      request = {
        ...request,
        signature: await signFreshCanonicalPayload(this.signingKey, payload),
      };
    }
    return this.http.postDirectoryAuth<Identity>(
      `/registry/names/${encodeURIComponent(name)}/transfer`,
      request,
    );
  }

  /**
   * Assign this name as the owner wallet's primary handle. Clears the primary
   * flag on the wallet's other names (one primary per wallet) and locks this
   * name from sale until it is unassigned.
   */
  async assignPrimary(name: string): Promise<Identity> {
    return this.setPrimary(name, true);
  }

  /**
   * Unassign this name as primary, leaving it unassigned and therefore sellable.
   */
  async unassignPrimary(name: string): Promise<Identity> {
    return this.setPrimary(name, false);
  }

  private async setPrimary(name: string, primary: boolean): Promise<Identity> {
    const action = primary ? "identity.assign" : "identity.unassign";
    const body: { signature?: string } = {};
    if (this.signingKey) {
      const payload = canonicalPayload(action, { username: name });
      body.signature = await signFreshCanonicalPayload(this.signingKey, payload);
    }
    return this.http.postDirectoryAuth<Identity>(
      `/registry/names/${encodeURIComponent(name)}/${primary ? "assign" : "unassign"}`,
      body,
    );
  }

  async claim(name: string, request: IdentityClaimRequest): Promise<Identity> {
    if (this.signingKey && !request.signature) {
      const payload = canonicalPayload("identity.claim", {
        cryptoId: request.cryptoId,
        publicKey: request.publicKey,
        username: name,
      });
      request = {
        ...request,
        signature: await signFreshCanonicalPayload(this.signingKey, payload),
      };
    }
    return this.http.post<Identity>(
      `/registry/names/${encodeURIComponent(name)}/claim`,
      request,
    );
  }

  async createSubname(
    name: string,
    request: SubnameCreateRequest,
  ): Promise<Subname> {
    if (this.signingKey && !request.signature) {
      const payload = canonicalPayload("identity.subname.create", {
        bio: request.bio,
        subname: request.subname,
        target: request.target,
        username: name,
      });
      request = {
        ...request,
        signature: await signFreshCanonicalPayload(this.signingKey, payload),
      };
    }
    return this.http.postDirectoryAuth<Subname>(
      `/registry/names/${encodeURIComponent(name)}/subnames`,
      request,
    );
  }

  async deleteSubname(name: string, subname: string): Promise<Identity> {
    const headers: Record<string, string> = {};
    if (this.signingKey) {
      const payload = canonicalPayload("identity.subname.delete", {
        subname,
        username: name,
      });
      headers["X-TinyPlace-Signature"] = await signFreshCanonicalPayload(
        this.signingKey,
        payload,
      );
      // Present the signing key so the backend can authorize a delegated hot
      // session key (the X-TinyPlace-Signature above is the ownership proof).
      const presentedKey = this.http.signingPublicKey();
      if (presentedKey) {
        headers["X-TinyPlace-Public-Key"] = presentedKey;
      }
    }
    return this.http.deletePublic<Identity>(
      `/registry/names/${encodeURIComponent(name)}/subnames/${encodeURIComponent(subname)}`,
      undefined,
      headers,
    );
  }

  private async registerRetryingPayment(
    request: RegisterRequest,
    options: {
      attempts?: number;
      intervalMs?: number;
      retryErrors?: Array<string>;
    },
  ): Promise<Identity> {
    const attempts = options.attempts ?? DEFAULT_REGISTRATION_ATTEMPTS;
    const intervalMs =
      options.intervalMs ?? DEFAULT_REGISTRATION_INTERVAL_MS;
    const retryErrors =
      options.retryErrors ?? DEFAULT_REGISTRATION_RETRY_ERRORS;
    let lastError: unknown;

    for (let attempt = 0; attempt < attempts; attempt += 1) {
      try {
        return await this.register(request);
      } catch (error) {
        lastError = error;
        if (!isRetryablePaymentError(error, retryErrors)) {
          throw error;
        }
        if (attempt + 1 < attempts && intervalMs > 0) {
          await sleep(intervalMs);
        }
      }
    }

    throw lastError;
  }
}

function normalizeHandle(name: string): string {
  const trimmed = name.trim();
  if (trimmed.startsWith("@")) {
    return trimmed;
  }
  return `@${trimmed}`;
}

function generateRegistrationNonce(username: string): string {
  const random = new Uint8Array(12);
  globalThis.crypto.getRandomValues(random);
  const suffix = Array.from(random, (byte) =>
    byte.toString(16).padStart(2, "0"),
  ).join("");
  return `register_${username.replace(/^@+/, "")}_${suffix}`;
}

function sleep(intervalMs: number): Promise<void> {
  return new Promise((resolve) => {
    setTimeout(resolve, intervalMs);
  });
}

function isRetryablePaymentError(
  error: unknown,
  retryErrors: Array<string>,
): boolean {
  if (!(error instanceof TinyVerseError) || error.status !== 402) {
    return false;
  }
  const message = paymentErrorMessage(error.body);
  return retryErrors.some((retryError) =>
    message.includes(retryError.toLowerCase()),
  );
}

function paymentErrorMessage(body: unknown): string {
  if (
    typeof body === "object" &&
    body !== null &&
    "error" in body &&
    typeof body.error === "string"
  ) {
    return body.error.toLowerCase();
  }
  if (typeof body === "string") {
    return body.toLowerCase();
  }
  return "";
}

function attachRegistrationPayment(
  error: unknown,
  payment: SolanaX402PaymentExecution | X402PaymentMap,
): unknown {
  if (typeof error === "object" && error !== null) {
    const failure = error as SolanaRegistrationFailure;
    failure.registrationPayment = payment;
    failure.onChainTx = registrationPaymentTx(payment);
  }
  return error;
}

function registrationPaymentTx(
  payment: SolanaX402PaymentExecution | X402PaymentMap,
): string | undefined {
  if (
    "payment" in payment &&
    "signature" in payment &&
    typeof payment.signature === "string"
  ) {
    return payment.signature;
  }
  const paymentMap = payment as X402PaymentMap;
  return paymentMap["onChainTx"] ?? paymentMap["tx"] ?? paymentMap["transaction"];
}

function registrationSignaturePayload(request: RegisterRequest): string {
  return JSON.stringify({
    action: "identity.register",
    fields: {
      // A handle is just a pointer now: registration binds the cryptoId to the
      // public key. Profile fields (bio/name/metadata) live on the wallet's
      // User and are set separately via UsersApi. `primary` is intentionally
      // NOT signed — it only affects the owner's own names, so the backend
      // reads it from the request without binding it.
      cryptoId: request.cryptoId,
      paymentMethods: request.paymentMethods
        ? request.paymentMethods.map(paymentMethodPayload)
        : null,
      publicKey: request.publicKey,
      username: request.username,
    },
  });
}

function paymentMethodPayload(method: PaymentMethod): Record<string, unknown> {
  return {
    network: method.network,
    address: method.address,
    assets: method.assets,
  };
}

File: sdk/python/src/tinyplace/api/registry.py
from __future__ import annotations

import asyncio
from typing import Any

from ..auth import sign_fresh_canonical_payload
from ..crypto import canonical_payload, crypto_id_to_public_key_base64
from ..http import HttpClient, TinyPlaceError, encode
from ..signer import Signer
from ..solana import (
    SOLANA_MAINNET_NETWORK,
    SOLANA_USDC_MINT,
    USDC_DECIMALS,
    execute_solana_x402_payment,
)
from ..types import Json, JsonDict

DEFAULT_REGISTRATION_ATTEMPTS = 30
DEFAULT_REGISTRATION_INTERVAL_MS = 3000
DEFAULT_REGISTRATION_RETRY_ERRORS = ["transaction not found", "insufficient confirmations"]


class RegistryApi:
    def __init__(self, http: HttpClient, signer: Signer | None = None) -> None:
        self._http = http
        self._signer = signer

    async def register(self, request: JsonDict) -> Json:
        request = _normalize_register_request(request)
        if self._signer and not request.get("signature"):
            request["signature"] = await sign_fresh_canonical_payload(
                self._signer,
                _registration_signature_payload(request),
            )
        return await self._http.post_public("/registry/names", request)

    async def register_with_solana_payment(
        self,
        request: JsonDict,
        *,
        rpc_url: str,
        secret_key: str | bytes,
        mint: str | None = None,
        decimals: int = USDC_DECIMALS,
        network: str | None = None,
        attempts: int = DEFAULT_REGISTRATION_ATTEMPTS,
        interval_ms: int = DEFAULT_REGISTRATION_INTERVAL_MS,
    ) -> JsonDict:
        """Register ``request``, settling the x402 fee with an on-chain Solana payment.

        Probes the registration to read the 402 payment challenge, executes the
        SPL/USDC (or native-SOL) transfer on chain, then retries the
        registration with the signed x402 payment map attached — polling through
        the brief window where the chain hasn't yet confirmed the transfer.
        Mirrors the TS SDK's ``registerWithSolanaPayment``.
        """
        if self._signer is None:
            raise ValueError("register_with_solana_payment requires a signer")
        normalized = _normalize_register_request(request)
        challenge = await self._registration_payment_challenge(normalized)
        amount = challenge.get("amount")
        recipient = challenge.get("to")
        if not amount or not recipient:
            raise ValueError("registration payment challenge is missing amount or recipient")

        execution = await execute_solana_x402_payment(
            signer=self._signer,
            rpc_url=rpc_url,
            secret_key=secret_key,
            # The registration fee is USDC: default to the USDC mint so a
            # challenge that names the asset by its SPL mint address (rather than
            # the literal "USDC" symbol) still settles. Callers override for
            # devnet / custom deployments.
            mint=mint or SOLANA_USDC_MINT,
            decimals=decimals,
            payment={
                "scheme": challenge.get("scheme", "exact"),
                "network": challenge.get("network") or network or SOLANA_MAINNET_NETWORK,
                "asset": challenge.get("asset") or "USDC",
                "amount": amount,
                "from": normalized.get("cryptoId") or self._signer.agent_id,
                "to": recipient,
                "nonce": challenge.get("nonce"),
                "expiresAt": challenge.get("expiresAt"),
                "metadata": {
                    **(challenge.get("metadata") or {}),
                    "identity": normalized["username"],
                    "purpose": "registration",
                },
                "publicKeyBase64": normalized.get("publicKey"),
            },
        )
        try:
            identity = await self._register_retrying_payment(
                {**normalized, "payment": execution["payment"]}, attempts, interval_ms
            )
        except TinyPlaceError as exc:
            # The payment is already on chain. A 5xx can still come back after
            # the identity was persisted, so check whether the handle now exists
            # before failing — re-driving the flow would otherwise risk paying
            # twice. Re-raise (with the payment attached) only if it truly wasn't
            # created.
            identity = await self._recover_registered_identity(normalized["username"], exc)
            if identity is None:
                raise
        return {
            "identity": identity,
            "payment": execution["payment"],
            "onChainTx": execution["signature"],
        }

    async def _recover_registered_identity(
        self, username: str, exc: TinyPlaceError
    ) -> Json | None:
        """Return the identity if a post-payment 5xx still created the handle."""
        if exc.status < 500:
            return None
        try:
            result = await self.get(username)
        except TinyPlaceError:
            return None
        if isinstance(result, dict) and result.get("available") is False:
            return result.get("identity")
        return None

    async def _registration_payment_challenge(self, request: JsonDict) -> dict[str, Any]:
        try:
            await self.register(request)
        except TinyPlaceError as exc:
            if exc.status == 402 and exc.payment_required is not None:
                return exc.payment_required.payment
            raise
        raise ValueError("registration did not return a payment challenge")

    async def _register_retrying_payment(
        self, request: JsonDict, attempts: int, interval_ms: int
    ) -> Json:
        attempts = max(1, attempts)
        for attempt in range(attempts):
            try:
                return await self.register(request)
            except TinyPlaceError as exc:
                if attempt == attempts - 1 or not _should_retry_registration(exc):
                    raise
            if interval_ms > 0:
                await asyncio.sleep(interval_ms / 1000)
        raise RuntimeError("unreachable: registration retry loop exhausted")

    async def get(self, name: str) -> Json:
        return await self._http.get(f"/registry/names/{encode(name)}")

    async def export(self, name: str) -> Json:
        return await self._http.get(f"/registry/names/{encode(name)}/export")

    async def update_profile_visibility(self, name: str, update: JsonDict) -> Json:
        if self._signer and not update.get("signature"):
            update = {
                **update,
                "signature": await sign_fresh_canonical_payload(
                    self._signer,
                    canonical_payload(
                        "identity.profile.visibility",
                        {
                            "activity": update.get("activity"),
                            "agentCard": update.get("agentCard"),
                            "attestations": update.get("attestations"),
                            "broadcasts": update.get("broadcasts"),
                            "groups": update.get("groups"),
                            "searchEngineIndexing": update.get("searchEngineIndexing"),
                            "username": name,
                        },
                    ),
                ),
            }
        return await self._http.put_directory_auth(
            f"/registry/names/{encode(name)}/profile-visibility",
            update,
        )

    async def renew(self, name: str, request: JsonDict) -> Json:
        if self._signer and not request.get("signature"):
            request = {
                **request,
                "signature": await sign_fresh_canonical_payload(
                    self._signer,
                    canonical_payload("identity.renew", {"username": name}),
                ),
            }
        return await self._http.post_directory_auth(
            f"/registry/names/{encode(name)}/renew",
            request,
        )

    async def transfer(self, name: str, request: JsonDict) -> Json:
        if self._signer and not request.get("signature"):
            request = {
                **request,
                "signature": await sign_fresh_canonical_payload(
                    self._signer,
                    canonical_payload(
                        "identity.transfer",
                        {
                            "cryptoId": request.get("cryptoId"),
                            "publicKey": request.get("publicKey"),
                            "username": name,
                        },
                    ),
                ),
            }
        return await self._http.post_directory_auth(
            f"/registry/names/{encode(name)}/transfer",
            request,
        )

    async def assign_primary(self, name: str) -> Json:
        return await self._set_primary(name, True)

    async def unassign_primary(self, name: str) -> Json:
        return await self._set_primary(name, False)

    async def claim(self, name: str, request: JsonDict) -> Json:
        if self._signer and not request.get("signature"):
            request = {
                **request,
                "signature": await sign_fresh_canonical_payload(
                    self._signer,
                    canonical_payload(
                        "identity.claim",
                        {
                            "cryptoId": request.get("cryptoId"),
                            "publicKey": request.get("publicKey"),
                            "username": name,
                        },
                    ),
                ),
            }
        return await self._http.post(f"/registry/names/{encode(name)}/claim", request)

    async def create_subname(self, name: str, request: JsonDict) -> Json:
        if self._signer and not request.get("signature"):
            request = {
                **request,
                "signature": await sign_fresh_canonical_payload(
                    self._signer,
                    canonical_payload(
                        "identity.subname.create",
                        {
                            "bio": request.get("bio"),
                            "subname": request.get("subname"),
                            "target": request.get("target"),
                            "username": name,
                        },
                    ),
                ),
            }
        return await self._http.post_directory_auth(
            f"/registry/names/{encode(name)}/subnames",
            request,
        )

    async def delete_subname(self, name: str, subname: str) -> Json:
        headers = {}
        if self._signer:
            headers["X-TinyPlace-Signature"] = await sign_fresh_canonical_payload(
                self._signer,
                canonical_payload("identity.subname.delete", {"subname": subname, "username": name}),
            )
            presented_key = self._http.signing_public_key()
            if presented_key:
                headers["X-TinyPlace-Public-Key"] = presented_key
        return await self._http.delete_public(
            f"/registry/names/{encode(name)}/subnames/{encode(subname)}",
            headers=headers,
        )

    async def _set_primary(self, name: str, primary: bool) -> Json:
        action = "identity.assign" if primary else "identity.unassign"
        body: dict[str, Any] = {}
        if self._signer:
            body["signature"] = await sign_fresh_canonical_payload(
                self._signer,
                canonical_payload(action, {"username": name}),
            )
        suffix = "assign" if primary else "unassign"
        return await self._http.post_directory_auth(f"/registry/names/{encode(name)}/{suffix}", body)


def _normalize_handle(username: str) -> str:
    return username if username.startswith("@") else f"@{username}"


def _should_retry_registration(exc: TinyPlaceError) -> bool:
    """True when a registration failed only because the payment isn't confirmed yet."""
    haystack = f"{exc} {exc.body}".lower()
    return any(error in haystack for error in DEFAULT_REGISTRATION_RETRY_ERRORS)


def _normalize_register_request(request: JsonDict) -> JsonDict:
    # Normalize the handle and derive the publicKey from cryptoId when omitted.
    # A Solana cryptoId IS the base58 ed25519 public key, so the base64 publicKey
    # the backend stores and signs over is derivable from it. Deriving here keeps
    # the signed payload and request body identical to the backend's server-side
    # derivation (so signatures verify) and lets callers pass just username +
    # cryptoId.
    normalized: JsonDict = {**request, "username": _normalize_handle(str(request["username"]))}
    crypto_id = normalized.get("cryptoId")
    if not normalized.get("publicKey") and crypto_id:
        normalized["publicKey"] = crypto_id_to_public_key_base64(str(crypto_id))
    return normalized


def _registration_signature_payload(request: JsonDict) -> str:
    # Must byte-match the backend's registrationPayload
    # (backend-tinyplace/internal/identity/auth.go): exactly these four fields,
    # no actorType/primary. Both sides serialize canonically (sorted keys, null
    # for absent values), so any extra signed field breaks verification (401).
    return canonical_payload(
        "identity.register",
        {
            "cryptoId": request.get("cryptoId"),
            "paymentMethods": request.get("paymentMethods"),
            "publicKey": request.get("publicKey"),
            "username": request.get("username"),
        },
    )
